### PR TITLE
Added a command to /eco that allows users to take percentages from pl…

### PR DESCRIPTION
**Pull request by CrashCringle12**
…ayers

Command syntax: /eco takeper <player> <amount>

Updated messages and command desc accordingly.

### CODE REVIEW

**Review by Ichbinjoe — CHANGES_REQUESTED** [review]
Why a new command? Why not just 'overload' take?

**Comment by CrashCringle12** [comment]
I think if take was hypothetically overloaded, give should also receive the same.

I figured taking percentages from someone's balance could quickly get real complicated for that edge case where you have to use the command multiple times. Accidentally forgetting the percent sign might cause problems and consequently as would not noticing. My thought process was have the second command so there's a clear distinction between the two and if the executor types the percent command it's more likely that it was intentional.

Hmm thoughts?

**Comment by pop4959** [comment]
In my opinion "takeper" just sounds silly (just go with takepercent if you must).

More seriously though, I don't think it's that easy to accidentally type the percent symbol. If you're typing in numbers, chances are you are not holding shift. Also, if you are adding the ability to take a percent, it would also make sense to include the functionality for give and set as well. As is, implementing it the way you are is duplicating a lot of code.

**Comment by Ichbinjoe** [comment]
Agreed - if one is merged so should the other.

I agree with @pop4959 - you are basically duplicating all functionality besides the number / value parsing. I don't believe that the additional % would result in mistakes - I would expect people are fairly diligent about being careful when shuffling virtual currency around. UI > safety in this instance.

**Comment by CrashCringle12** [comment]
Hmm alright those are actually some good points. I'll make the change

**Comment by CrashCringle12** [comment]
Hm so I removed the new command and condensed things as much as possible. I added one method to handle percentages with take/give. 

I'm unsure why all of my builds are failing tho, I haven't modified anything aside from these same files, any idea what's happening here?

**Comment by pop4959** [comment]
It's not your fault, Spigot added some new methods that is causing Essentials tests to fail:
```
[ERROR] /home/runner/work/Essentials/Essentials/Essentials/test/com/earth2me/essentials/FakeServer.java:[517,47] <anonymous com.earth2me.essentials.FakeServer$2> is not abstract and does not override abstract method setStatistic(org.bukkit.Statistic,org.bukkit.entity.EntityType,int) in org.bukkit.OfflinePlayer
```

I'll force a rebuild on this PR later so you can see the real status.

**Comment by CrashCringle12** [comment]
Ahhh I see, ty ty.

**Comment by pop4959** [comment]
The build should pass once #3140 is merged.

**Comment by JRoy** [comment]
(And once this PR rebased)

**Comment by pop4959** [comment]
Actually, pretty sure the diff from this PR should still work. He might need to rebase if he isn't able to build on his side, but chances are he's using a cached version of the Spigot API. If he runs maven clean and/or invalidates cache in IntelliJ he may have to do that though.

**Comment by CrashCringle12** [comment]
Alright this should be good, let me know what you guys think. It works for set, give, and take with its own message for each.
And i didn't feel like waiting for the pull request (lol) so i just updated FakeServer to get the builds working.

File: Essentials/src/com/earth2me/essentials/User.java [hunk 1]
```diff
@@ -220,7 +220,34 @@ public void takeMoney(final BigDecimal value, final CommandSource initiator, Use
...
     }
-
+    public void percentHandler(final BigDecimal percent, final CommandSource initiator, UserBalanceUpdateEvent.Cause cause, int cmd) {
+        String message = "PercentGivenTo";
+        if (percent.signum() == 0) {
+            return;
+        }
+        BigDecimal value = getMoney().multiply(percent.divide(new BigDecimal("100.0")));
+        try {
+            switch (cmd) {
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/User.java]: Noooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooope

Begone with your magic numbers!!!!

Make this an enum or refactor out - prefer the second to the first but :man_shrugging:

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 2]
```diff
@@ -35,7 +36,13 @@ public void run(final Server server, final CommandSource sender, final String co
 
         try {
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
-            amount = (cmd == Commandeco.EcoCommands.RESET) ? startingBalance : new BigDecimal(args[2].replaceAll("[^0-9\\.]", ""));
+            if (args[2].contains("%") && !(cmd == Commandeco.EcoCommands.RESET)) {
+                isPercent = true;
+                amount = new BigDecimal(args[2].substring(0,args[2].length()-1));
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: This is not sane - this will pass 5%3, and will then throw an exception because it can't parse the number.... which is then sent back to the user as 'NotEnoughArgumentsException'.

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 3]
```diff
@@ -55,13 +62,15 @@ public void run(final Server server, final CommandSource sender, final String co
     protected void updatePlayer(final Server server, final CommandSource sender, final User player, final String[] args) throws NotEnoughArgumentsException, ChargeException, MaxMoneyException {
         switch (cmd) {
             case GIVE:
-                player.giveMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
+                if (isPercent)
+                    player.percentHandler(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO, 0);
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: Please refactor into something along the lines of 'giveMoneyPercent' instead of using these magic values and a switch.

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 4]
```diff
@@ -91,10 +104,14 @@ private void set(BigDecimal amount, final User player, final CommandSource sende
         BigDecimal maxBalance = ess.getSettings().getMaxMoney();
         boolean underMinimum = (amount.compareTo(minBalance) < 0);
         boolean aboveMax = (amount.compareTo(maxBalance) > 0);
-        player.setMoney(underMinimum ? minBalance : aboveMax ? maxBalance : amount, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
-        player.sendMessage(tl("setBal", NumberUtil.displayCurrency(player.getMoney(), ess)));
-        if (sender != null) {
-            sender.sendMessage(tl("setBalOthers", player.getDisplayName(), NumberUtil.displayCurrency(player.getMoney(), ess)));
+        if (isPercent) {
+            player.percentHandler(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO, 2);
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: This implementation doesn't line up with the implementation below... I get its non-congruent with the above two methods but this way makes it look like the messages are missing.

**Review by Ichbinjoe — CHANGES_REQUESTED** [review]
I think I see what you were trying to achieve - I think it would be simpler if there were simple forks in the code which that when a percentage was encountered would reset amount to the proportional amount (depending on context) and continue on the original execution - I don't think any value is lost by not telling them the particular percentage.

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java (hunk 2 above)
Comment (CrashCringle12) [reply Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: Oof alright alright, fair. That is pretty bonkers now that you mention it. I'll check if the last character is '%' rather than if the entire string contains one.

```suggestion
if (args[2].endsWith("%") && (cmd != Commandeco.EcoCommands.RESET))
```

**Comment by CrashCringle12** [comment]
Hm alright, so where it stands now percentages are handled properly and everything extra except for the messages I added have been yeeted.

It feels like percentage should be displayed though. Whenever I use the command there's a lack of confirmation I guess? I mean it tells me the balance taken away sure, but I think seeing the percent would be best.

I have a number of different ideas for how to make it work, but I'd be taking up a similar method to that of my original commit (minus the extra command ofc). Am I better off making additional methods in User or within Eco? As I'd be making 3 mostly copied new methods or 1 method that uses an ENUM to tell whether it's take, give, or set. Or i mean i could modify take and add a give, and set method to Eco. Just sort of brain dumping the ideas i have rn before I make another attempt at it. Unless you guys think it's really fine without the message

**Comment by Ichbinjoe** [comment]
@pop4959 for opinion about message.

New code should follow the convention of the code which already exists. Enum to define behavior is usually a code smell (especially since we don't have a good reason to really use it).

**Comment by pop4959** [comment]
As long as there is already a confirmation for the amount removed/added to the account, I don't think it's necessary to add new messages specifically for this. You could just calculate the final amount based on the percent and let the existing code handle it. It shouldn't be too confusing for users since they can see what the current balance is, as well as what was removed from it.

**Comment by CrashCringle12** [comment]
Hmm I suppose, though I think it's worth mentioning tho you don't know the previous balance beforehand unless you check.
image
I definitely see where you're saying tho and understand, I mean it makes sense... just I always operate on the notion that people can always be numskulls lmao.

I removed the messages and just left the percent functionality. I'm alright compromising on that. How am I looking now?

**Comment by pop4959** [comment]
Well, people can be stupid any way you cut it. Since the % thing is kind of an easter egg here, it shouldn't be much of a problem anyways. It's also hard to accidentally type, as aforementioned.

Also you actually do know the prior balance.

New balance = Starting balance - Balance removed (e.g. $900 = $1000 - $100 (10%))
->
Starting balance = New balance + Balance removed (e.g. $1000 = $900 + $100)

Simple arithmetic that can be worked out, if you make a mistake. Knowing the percent you typed in isn't even necessary, but you could use chat history to check that as well.

File: Essentials/src/com/earth2me/essentials/User.java [hunk 5]
```diff
@@ -220,7 +220,7 @@ public void takeMoney(final BigDecimal value, final CommandSource initiator, Use
             initiator.sendMessage(tl("takenFromOthersAccount", NumberUtil.displayCurrency(value, ess), this.getDisplayName(), NumberUtil.displayCurrency(getMoney(), ess)));
         }
     }
-
+
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/User.java]: Try to avoid unnecessary formatting changes, as this results in a dirty diff.

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 6]
```diff
@@ -35,7 +36,9 @@ public void run(final Server server, final CommandSource sender, final String co
 
         try {
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
+            isPercent = (args[2].endsWith("%"));
             amount = (cmd == Commandeco.EcoCommands.RESET) ? startingBalance : new BigDecimal(args[2].replaceAll("[^0-9\\.]", ""));
+
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: =

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 7]
```diff
@@ -61,7 +67,7 @@ protected void updatePlayer(final Server server, final CommandSource sender, fin
             case TAKE:
                 take(amount, player, sender);
                 break;
-
+
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: =

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 8]
```diff
@@ -73,7 +79,7 @@ private void take(BigDecimal amount, final User player, final CommandSource send
         BigDecimal money = player.getMoney();
         BigDecimal minBalance = ess.getSettings().getMinMoney();
         if (money.subtract(amount).compareTo(minBalance) >= 0) {
-            player.takeMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
+                player.takeMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: =

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 9]
```diff
@@ -91,12 +97,12 @@ private void set(BigDecimal amount, final User player, final CommandSource sende
         BigDecimal maxBalance = ess.getSettings().getMaxMoney();
         boolean underMinimum = (amount.compareTo(minBalance) < 0);
         boolean aboveMax = (amount.compareTo(maxBalance) > 0);
-        player.setMoney(underMinimum ? minBalance : aboveMax ? maxBalance : amount, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
-        player.sendMessage(tl("setBal", NumberUtil.displayCurrency(player.getMoney(), ess)));
-        if (sender != null) {
-            sender.sendMessage(tl("setBalOthers", player.getDisplayName(), NumberUtil.displayCurrency(player.getMoney(), ess)));
+            player.setMoney(underMinimum ? minBalance : aboveMax ? maxBalance : amount, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: = (all of this under this line)

File: Essentials/src/messages.properties [hunk 10]
```diff
@@ -379,7 +379,7 @@ onlyPlayerSkulls=\u00a74You can only set the owner of player skulls (\u00a7c397\
 onlySunStorm=\u00a74/weather only supports sun/storm.
 openingDisposal=\u00a76Opening disposal menu...
 orderBalances=\u00a76Ordering balances of\u00a7c {0} \u00a76users, please wait...
-oversizedMute=�4You may not mute a player for this period of time.
+oversizedMute=\u00a74You may not mute a player for this period of time.
```
Comment (pop4959) [inline Essentials/src/messages.properties]: This isn't relevant to this PR. If there is an issue with this message, please open another PR.

**Review by pop4959 — CHANGES_REQUESTED** [review]

**Comment by pop4959** [comment]
Just clean up this PR and show that you've tested the commands and we're probably good.

**Comment by CrashCringle12** [comment]
image
image
image

**Comment by CrashCringle12** [comment]
Alright how we looking

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 11]
```diff
@@ -20,7 +20,7 @@
 public class Commandeco extends EssentialsLoopCommand {
     Commandeco.EcoCommands cmd;
     BigDecimal amount;
-
+	boolean isPercent;
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: This should be 4 spaces to remain consistent with the code formatting in this file.

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 12]
```diff
@@ -35,6 +35,7 @@ public void run(final Server server, final CommandSource sender, final String co
 
         try {
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
+            isPercent = (args[2].endsWith("%"));
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: Don't need parentheses here.

**Review by pop4959 — CHANGES_REQUESTED** [review]

**Comment by pop4959** [comment]
@CrashCringle12 Sorry for the long wait, forgot to come back to this issue. Minor changes to be made, but the diff is way wayyy cleaner than last time, overall. I'm not really a big fan of global variables floating around personally, but since the command seems to use them anyway for the command type and amount, I don't see the harm.

@Ichbinjoe How're you feeling about this?

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 13]
```diff
@@ -53,6 +54,9 @@ public void run(final Server server, final CommandSource sender, final String co
 
     @Override
     protected void updatePlayer(final Server server, final CommandSource sender, final User player, final String[] args) throws NotEnoughArgumentsException, ChargeException, MaxMoneyException {
+        if(isPercent && cmd != EcoCommands.RESET) {
```
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: Forgot, there should also be a space between the `if` and `(` here. By the way, using CTRL+ALT+L in IntelliJ will auto format the file for you, in standard formatting (which is what is used here).

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java [hunk 14]
```diff
@@ -53,6 +54,9 @@ public void run(final Server server, final CommandSource sender, final String co
...
+            amount = player.getMoney().multiply(amount.divide(new BigDecimal("100.0")));
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: This can be optimized quite a bit.
```suggestion
            amount = player.getMoney().multiply(amount).scaleByPowerOfTen(-2);
```

This abuses how BigDecimals & Integers work to get around a division.

**Review by Ichbinjoe — CHANGES_REQUESTED** [review]

File: Essentials/src/com/earth2me/essentials/commands/Commandeco.java (hunk 11 above)
Comment (CrashCringle12) [reply Essentials/src/com/earth2me/essentials/commands/Commandeco.java]: Gotcha

**Comment by CrashCringle12** [comment]
You're all good, take your time. Took me a while to respond myself.
I fixed the changes you guys suggested !

**Review by Ichbinjoe — APPROVED** [review]

**Review by pop4959 — APPROVED** [review]
